File: Lab_4/omega.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>


int main(int argc, char* argv[])
{
	short intrari, conexiuni;
	int i;
	int aux;

	printf("Numar intrari = ");
	scanf("%hd", &intrari);
	printf("Numar conexiuni = ");
	scanf("%hd", &conexiuni);

	short in[conexiuni], out[conexiuni];

	for (i = 0; i < conexiuni; ++i)
	{
		printf("In_%d = ", i+1);
		scanf("%hd", &in[i]);
		printf("Out_%d = ", i+1);
		scanf("%hd", &out[i]);
	}



	short etaje = log((double)intrari) / log(2.0);
	short mat[intrari][etaje+1];
	memset(mat, 0, sizeof(short) * intrari * (etaje+1));
	int j;

	printf("Etaje %hu\n", etaje);
	printf("Rezolvare \n");

	j = 0;
	while(j < conexiuni){
		short xor_tag = in[j] ^ out[j];

		printf("\n - Din %hu in %hu - \n", in[j], out[j]);

		if (in[j] >= 4)
		{
			in[j] = in[j] << 1;
			in[j] = in[j] + 1;
			in[j] = in[j] & 7;
		}
		else
		{
			in[j] = in[j] << 1;
		}
		if(mat[in[j]][0])
		{
			printf("Conflict\n");
		}
		else
		{

			printf("Mat[%d][%d] = %d\n", in[j], 0, mat[in[j]][i+1]);
			mat[in[j]][0] = 1;

			for(i = 0;  i < etaje; i++)
			{
				/* Verific bitul coresp stage-ului */
				aux = xor_tag & (1 << (etaje-i-1));

				printf("Comutatorul de pe etaj %hd - ", etaje-i-1);

				/* Direct */
				if (aux == 0)
				{
					/* Unde trebuie sa rutez mai departe */
					if (in[j] >= 4)
					{
						in[j] = in[j] << 1;
						in[j] = in[j] + 1;
						in[j] = in[j] & 7;
					}
					else
					{
						in[j] = in[j] << 1;
					}
					if(mat[in[j]][i+1])
					{
						printf("Conflict\n");
						break;
					}
					else
					{
						mat[in[j]][i+1] = 1;
						printf("Direct (C0 = 0; C1 = 0)\n");
					}

				}
				/* Cross */
				else
				{
					/* Este cross deci se modifica cu +1 sau -1 */
					if (in[j] % 2 == 0)
						++in[j];
					else
						--in[j];

					/* Unde trebuie sa rutez mai departe */
					if (in[j] >= 4)
					{
						in[j] = in[j] << 1;
						in[j] = in[j] + 1;
						in[j] = in[j] & 7;
					}
					else
					{
						in[j] = in[j] << 1;
					}
					if(mat[in[j]][i+1])
					{
						printf("Conflict\n");
						break;
					}
					else
					{
						mat[in[j]][i+1] = 1;
						printf("Invers (C0 = 1; C1 = 1)\n");
					}
				}
			}
		}
		j++;
	}
	return 0;
}

File: Lab_4/omega.py
from sys import stdin
from math import *

print "No entrys:",
entry = int(stdin.readline())
print "No connections:",
connections = int(stdin.readline())

inputs = [];
outputs = [];

for i in xrange(0 , connections):
	print "Input %d:" % (i),
	inputs.append(int(stdin.readline()));
	print "Output %d:" % (i),
	outputs.append(int(stdin.readline()));

stages = int(log(entry,2));
mat = [[0 for y in range(stages+1)] for x in range(entry)];

print "Computing...";

for i in xrange(0, connections):

	xor_tag = inputs[i] ^ outputs[i];

	print"Input %d to output %d" % (inputs[i], outputs[i])

	if(inputs[i] < 4):
		inputs[i] <<= 1
	else:
		inputs[i] <<= 1;
		inputs[i] +=  1;
		inputs[i] &=  7;
	if(mat[inputs[i]][0]):
		print "Conflict";
	else:
		# print "Mat[%d][%d] = %d" % (inputs[i],0,mat[inputs[i]][0])
		mat[inputs[i]][0] = 1;
		for j in xrange(0, stages):
			comut = xor_tag & (1 << (stages-j-1))

			print "Switch on stage %d - " % (stages-j-1),

			# Direct
			if (comut == 0):
				# Unde trebuie sa rutez
				if(inputs[i] < 4):
					inputs[i] <<= 1
				else:
					inputs[i] <<= 1;
					inputs[i] +=  1;
					inputs[i] &=  7;
				if(mat[inputs[i]][j+1]):
					print "Conflict";
					break;
				else:
					# print "Mat[%d][%d] = %d" % (inputs[i],j+1,mat[inputs[i]][j+1])
					mat[inputs[i]][j+1] = 1;
					print "Direct (C0 = 0; C1 = 0)"
			# Cross
			else:
				# Verific daca fac cross in sus sau in jos
				if (inputs[i] % 2 == 0):
					inputs[i]+=1;
				else:
					inputs[i]-=1;

				#  Unde trebuie sa rutez
				if(inputs[i] < 4):
					inputs[i] <<= 1
				else:
					inputs[i] <<= 1;
					inputs[i] +=  1;
					inputs[i] &=  7;

				if(mat[inputs[i]][j+1]):
					print "Conflict";
					break;
				else:
					# print "Mat[%d][%d] = %d" % (inputs[i],j+1,mat[inputs[i]][j+1])
					mat[inputs[i]][j+1] = 1;
					print "Invers (C0 = 1; C1 = 1)"
